### PR TITLE
[CALCITE] Provide column-level attributes in addition to type information

### PR DESCRIPTION
**Pull request by AndrewPochapsky**
Added support for a RelDataType to have associated column attributes

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java [hunk 1]
```diff
@@ -94,12 +142,27 @@ public BasicSqlType(
       int scale,
       SqlCollation collation,
       SerializableCharset wrappedCharset) {
+    this(typeSystem, typeName, nullable, precision, scale, collation, wrappedCharset,
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java:145]: nit: Line wrap

File: core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java [hunk 2]
```diff
@@ -191,6 +259,10 @@ protected void generateTypeString(StringBuilder sb, boolean withDetail) {
     if (!withDetail) {
       return;
     }
+    for (SqlColumnAttribute attribute : attributes) {
+      sb.append(" ").append(writer.format(attribute));
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java:263]: nit: Use append with single quotes for a single character.

File: core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactory.java [hunk 3]
```diff
@@ -209,6 +210,18 @@ RelDataType createTypeWithCharsetAndCollation(
    */
   RelDataType createSqlType(SqlTypeName typeName);
 
+  /**
+   * Creates a SQL type with no precision or scale and the given column
+   * attributes.
+   *
+   * @param typeName Name of the type, for example {@link SqlTypeName#BOOLEAN},
+   *   never null
```
Comment (lisayin) [inline core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactory.java:218]: nit indentation

File: core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactory.java [hunk 4]
```diff
@@ -251,6 +283,27 @@ RelDataType createSqlType(
...
 
+  /**
+   * Creates a SQL type with precision, scale and column attributes.
+   *
+   * @param typeName  Name of the type, for example {@link SqlTypeName#DECIMAL}.
+   *                  Never null.
+   * @param precision Precision of the value.
+   *                  Must be non-negative or
+   *                  {@link RelDataType#PRECISION_NOT_SPECIFIED}.
+   * @param scale     scale of the values, i.e. the number of decimal places to
```
Comment (lisayin) [inline core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactory.java:294]: nit: capitalize "scale of the values"

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 5]
```diff
@@ -3440,7 +3440,9 @@ public void validateIntervalQualifier(SqlIntervalQualifier qualifier) {
       }
       SqlBasicTypeNameSpec basicTypeNameSpec =
           (SqlBasicTypeNameSpec) typeNameSpec;
-      builder.add(col.name.toString(), basicTypeNameSpec.sqlTypeName);
+      RelDataType type = typeFactory.createSqlType(basicTypeNameSpec.sqlTypeName,
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:3443]: nit: Line wrap

File: core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java [hunk 6]
```diff
@@ -71,6 +90,21 @@ public BasicSqlType(RelDataTypeSystem typeSystem, SqlTypeName typeName,
...
+   * Constructs a type with precision/length and attributes but no scale.
+   *
+   * @param typeSystem Type system
+   * @param typeName Type name
+   * @param precision Precision (called length for some types)
+   * @param attributes The column attributes, may not be null
+   */
+  public BasicSqlType(RelDataTypeSystem typeSystem, SqlTypeName typeName,
+      int precision, List<SqlColumnAttribute> attributes) {
+    this(typeSystem, typeName, false, precision, SCALE_NOT_SPECIFIED, null,
```
Comment (lisayin) [inline core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java:103]: nit: add inline comments for the bool and null parameters

**Pull request merged**